### PR TITLE
feat(db): prune lite node data automatically

### PR DESCRIPTION
**Pull request by lurais**
**What does this PR do?**
Prune lite node data automatically
**Why are these changes required?**

**This PR has been tested by:**
- Unit Tests
- Manual Testing

**Follow up**

**Extra details**

### CODE REVIEW

File: chainbase/src/main/java/org/tron/core/db2/core/SnapshotRoot.java [hunk 1]
```diff
@@ -38,6 +39,7 @@ public SnapshotRoot(DB<byte[], byte[]> db) {
     if (CACHE_DBS.contains(this.db.getDbName())) {
       this.cache = CacheManager.allocate(CacheType.findByType(this.db.getDbName()));
     }
+    ChainDataPruner.register(this.db.getDbName(), this);
```
[the remark this answers is not in the document]
Comment (lurais) [reply chainbase/src/main/java/org/tron/core/db2/core/SnapshotRoot.java:42]: In fact it has no effect.

File: chainbase/src/main/java/org/tron/common/storage/prune/ChainDataPruner.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+@Slf4j(topic = "db")
+@Component
+public class ChainDataPruner {
+
+  private static final String BLOCK_INDEX_STORE_NAME = "block-index";
+  private static final String BLOCK_STORE_NAME = "block";
+  private static final String TRANSACTION_STORE_NAME = "trans";
+  private static final String TRANSACTION_RET_STORE_NAME = "transactionRetStore";
+  private long BLOCKS_TO_RETAIN = 65536;
```
[the remark this answers is not in the document]
Comment (lurais) [reply chainbase/src/main/java/org/tron/common/storage/prune/ChainDataPruner.java:32]: ok

File: framework/src/test/java/org/tron/common/storage/prune/ChainDataPrunerTest.java [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+      blockStore.put(blockCapsule4.getBlockId().getBytes(), blockCapsule4);
+      blockStore.put(blockCapsule5.getBlockId().getBytes(), blockCapsule5);
+      blockIndexStore.put(blockCapsule1.getBlockId());
+      blockIndexStore.put(blockCapsule2.getBlockId());
+      blockIndexStore.put(blockCapsule3.getBlockId());
+      blockIndexStore.put(blockCapsule4.getBlockId());
+      blockIndexStore.put(blockCapsule5.getBlockId());
+      transactionRetStore.put(Longs.toByteArray(blockCapsule2.getNum()), transactionRetCapsule);
+      dynamicPropertiesStore.saveLatestBlockHeaderNumber(5L);
+      ReflectUtils.setFieldValue(chainDataPruner,"blocksToRetain",3);
```
Comment (halibobo1205) [inline framework/src/test/java/org/tron/common/storage/prune/ChainDataPrunerTest.java:160]: why by reflect

File: framework/src/main/java/org/tron/core/net/message/handshake/HelloMessage.java [hunk 4]
```diff
@@ -59,14 +61,33 @@ public HelloMessage(Node from, long timestamp, ChainBaseManager chainBaseManager
...
     this.type = MessageTypes.P2P_HELLO.asByte();
     this.data = this.helloMessage.toByteArray();
   }
 
+  public long calculateLowestBlockNum(ChainBaseManager chainBaseManager) {
+    if (!CommonParameter.getInstance().getStorage().isDbAutoPrune()) {
+      return chainBaseManager.isLiteNode()
+          ? chainBaseManager.getLowestBlockNum() : 0;
+    }
+    if (!chainBaseManager.isLiteNode()) {
```
Comment (halibobo1205) [inline framework/src/main/java/org/tron/core/net/message/handshake/HelloMessage.java:76]: check lite first before autoPrune

File: framework/src/main/java/org/tron/core/net/message/handshake/HelloMessage.java [hunk 5]
```diff
@@ -59,14 +61,33 @@ public HelloMessage(Node from, long timestamp, ChainBaseManager chainBaseManager
...
+    long latestBlockNum = chainBaseManager.getHeadBlockNum();
+    if (latestBlockNum - lowestBlockNum
+        <= CommonParameter.getInstance().getStorage().getDbAutoPruneRetain()
+        - 2 * Commons.ONE_DAY_BLOCKS_PREDICT) {
+      return lowestBlockNum;
+    } else {
+      return latestBlockNum - CommonParameter.getInstance().getStorage().getDbAutoPruneRetain()
+          + 2 * Commons.ONE_DAY_BLOCKS_PREDICT;
+    }
+  }
```
Comment (halibobo1205) [inline framework/src/main/java/org/tron/core/net/message/handshake/HelloMessage.java:89]: refactoring code for easy understanding

File: framework/src/main/java/org/tron/core/config/args/Args.java [hunk 6]
```diff
@@ -522,6 +522,11 @@ public static void setParam(final String[] args, final String confFileName) {
     PARAMETER.storage.setEstimatedBlockTransactions(
         Storage.getEstimatedTransactionsFromConfig(config));
     PARAMETER.storage.setMaxFlushCount(Storage.getSnapshotMaxFlushCountFromConfig(config));
+    PARAMETER.storage.setDbAutoPrune(Storage.getDbAutoPruneSwitchFromConfig(config));
+    PARAMETER.storage.setDbAutoPruneRetain(Storage.getDbAutoPruneRetainFromConfig(config));
+    PARAMETER.storage.setDbAutoPruneFrequency(
+        Storage.getDbAutoPruneFrequencyFromConfig(config));
```
Comment (halibobo1205) [inline framework/src/main/java/org/tron/core/config/args/Args.java:528]: deal  it in storage

File: chainbase/src/main/java/org/tron/core/db2/core/SnapshotRoot.java [hunk 7]
```diff
@@ -30,6 +33,8 @@ public class SnapshotRoot extends AbstractSnapshot<byte[], byte[]> {
   private TronCache<WrappedByteArray, WrappedByteArray> cache;
   private static final List<String> CACHE_DBS = CommonParameter.getInstance()
       .getStorage().getCacheDbs();
+  private static final Set<String> PRUNE_DBS = Arrays.asList( "block-index","block", "trans",
```
Comment (halibobo1205) [inline chainbase/src/main/java/org/tron/core/db2/core/SnapshotRoot.java:36]: move to ChainDataPruner

File: chainbase/src/main/java/org/tron/common/storage/prune/ChainDataPruner.java [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    long latestBlockNumber = chainBaseManager.getDynamicPropertiesStore().
+        getLatestBlockHeaderNumberFromDB();
+    if (latestBlockNumber - lowestBlockNumber + 1 > blocksToRetain) {
+      doPrune(lowestBlockNumber, latestBlockNumber);
+    }
+  }
+
+  private void doPrune(long lowestBlockNumber, long latestBlockNumber) {
+    long toFetchCount =
+        Math.min((latestBlockNumber - lowestBlockNumber + 1 - blocksToRetain), blocksBatchFlush);
```
Comment (halibobo1205) [inline chainbase/src/main/java/org/tron/common/storage/prune/ChainDataPruner.java:82]: performance？

**Review by halibobo1205 — CHANGES_REQUESTED** [review]

File: framework/src/main/java/org/tron/core/net/message/handshake/HelloMessage.java (hunk 5 above)
Comment (halibobo1205) [reply framework/src/main/java/org/tron/core/net/message/handshake/HelloMessage.java:87]: return Math.Max(latestBlockNum - retainBlocksSafeIn2Days + 1, lowestBlockNum)?

**Review by halibobo1205 — CHANGES_REQUESTED** [review]

**Review by 317787106 — APPROVED** [review]

**Review by halibobo1205 — APPROVED** [review]

**Review by guoquanwu — APPROVED** [review]

**Pull request merged**